plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java

```diff
@@ -243,26 +244,30 @@ public void finalizeSnapshot(SnapshotId snapshotId, ShardGenerations shardGenera
                                  Function<ClusterState, ClusterState> stateTransformer,
                                  ActionListener<Tuple<RepositoryData, SnapshotInfo>> listener) {
         if (SnapshotsService.useShardGenerations(repositoryMetaVersion) == false) {
-            listener = delayedListener(listener);
+            listener = delayedListener(listener, stateTransformer);
```

Comment:
This was broken, before. We must delay the final CS update as that's what keeps us safe on S3 without shard generations